### ISSUE TITLE
Does not work on Windows

### ISSUE THREAD

**Issue by onikienko** [extend-chrome/rollup-plugin-chrome-extension#137]
### Build tool

Vite

### Where do you see the problem?

- [ ] In the browser
- [X] In the terminal

### Describe the bug

`dev` and `build` scripts just hang on the `Building Chrome Extension files...` It appears on Windows only. When running on Linux everything works as expected.

### Reproduction

Create a new extension using this guide https://dev.to/jacksteamdev/create-a-vite-react-chrome-extension-in-90-seconds-3df7
Run it on Windows.

### Logs

```shell
$ npm run dev  
vite v2.7.3 dev server running at:

  > Local: http://localhost:3000/
  > Network: use `--host` to expose

  ready in 606ms.

Building Chrome Extension files...

// and script just hangs here eating CPU
```

### System Info

```shell
System:
    OS: Windows 11
    CPU: (16) x64 AMD Ryzen 7 5800X 8-Core Processor             
    Memory: 17.17 GB / 31.91 GB
  Binaries:
    Node: 16.13.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.15 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 8.1.0 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Chromium (96.0.1054.53)
  npmPackages:
    rollup-plugin-chrome-extension: ^4.0.1-9 => 4.0.1-9 
    vite: ^2.7.3 => 2.7.3
```

### Severity

blocking all usage of RPCE

**Comment by onikienko** [comment]
Also, I tried to run RPCE tests on Windows. Looks like there are problems with paths. Take a look at the screenshot. (Project location is C:\Projects\experiments\rollup-plugin-chrome-extension )

![image](https://user-images.githubusercontent.com/3234928/146431108-73a79d07-f16b-49b9-9038-a1f8699f605f.png)

**Comment by jacksteamdev** [comment]
@onikienko Those watch file paths look suspicious. Would you have time to help debug this? :pray: 

This _may_ be the offending code:
https://github.com/extend-chrome/rollup-plugin-chrome-extension/blob/fdf47a3ee891b1b1e7c4f76eb710f58d500e843f/src/files_spawnFile.ts#L65-L67

I added this to support Windows, but Rollup is [supposed to use POSIX separators internally] (https://github.com/rollup/rollup/issues/3993#issuecomment-797874991).

**Comment by jacksteamdev** [comment]
A quick tip: The tests take a long time to run altogether, so it might be helpful to debug just one test.

I use this VSCode plugin to debug single tests: [Jest Runner - Visual Studio Marketplace](https://marketplace.visualstudio.com/items?itemName=firsttris.vscode-jest-runner). It adds a CodeLens to the test file that runs a single test:
![image](https://user-images.githubusercontent.com/23390212/146436446-2ddf3c62-12c7-40ad-b87e-c0ffe6669d7a.png)

It may require adding some lines to the VSCode settings file: `.vscode/settings.json`
```
{
  "jestrunner.debugOptions": {
    "args": ["--no-cache"]
  },
  "jestrunner.jestPath": "node_modules/jest/bin/jest.js",
}
```

### XState Viz

When you run the debugger, it should open the XState Visualizer. You can see the progress of the build there. The dropdown lets you select individual files. The paths of the files should be available in the sidebar under "context"

![image](https://user-images.githubusercontent.com/23390212/146441963-407296f1-af43-44de-b941-fc2b01be95f2.png)

**Comment by onikienko** [comment]
Tried to debug some tests. Get this jest config error:
![image](https://user-images.githubusercontent.com/3234928/146578857-97780a43-bf0d-4362-8a31-0820f9c1457e.png)

In the jest config `moduleNameMapper` it looks like correct.
![image](https://user-images.githubusercontent.com/3234928/146579520-995a88a3-0b12-4898-93a2-7eb712bae76d.png)

Do you have an idea how to debug/fix it?

**Comment by jacksteamdev** [comment]
I wonder if we need to use the `path` module for the keys and handle both types of path separators in the regex?

**Comment by jacksteamdev** [comment]
`test/jest.globalSetup.ts` is supposed to run before Jest runs. It builds the files in `src/browser` and `src/service-worker`. It looks like something is wrong there. Do the files below exist? If not, then something is wrong in the global setup file.

![image](https://user-images.githubusercontent.com/23390212/146620090-d919b622-4683-43c6-a028-87df7ef2c9a2.png)

**Comment by onikienko** [comment]
You are right. It comes from `test/jest.globalSetup.ts`. Problem with rollup path resolution. Take a look at `source`. Slashes are missing here:

![image](https://user-images.githubusercontent.com/3234928/146642823-f1408e52-dd85-466e-a559-49c2c63b347a.png)

This one is even more complicated:

![image](https://user-images.githubusercontent.com/3234928/146642904-62beb766-6d3b-4738-9623-a4f1299e6987.png)

If honest, I have no idea how rollup works 😺 Looks like that `resolveId` is the first hook called. Is there a way to set up how paths should be resolved?

**Comment by jacksteamdev** [comment]
Rollup configs always accept POSIX, so let's try mapping `entryFiles` in line 31 to POSIX before giving it to Rollup in `config` in line 33. `normalizePath` from [@rollup/pluginutils](https://github.com/rollup/plugins/tree/master/packages/pluginutils#normalizepath) seems like a good place to start.

**Comment by onikienko** [comment]
Tried to `normalizePath` all paths for the rollup config. Paths look correct after that. But now rollup configs just hangs on bundleImports.  Reported https://github.com/jacksteamdev/rollup-plugin-bundle-imports/issues/13

Looks like the issue is out of my expertise 😅

**Comment by jacksteamdev** [comment]
Thanks for working on this! I'm going to remove `bundleImports` and just use a custom plugin.

Looks like `normalizePath` is the way forward. I'll add it everywhere I can; maybe you can give it a run then?

**Comment by jacksteamdev** [comment]
@onikienko Can you make a PR with the changes you've made?

You've done a lot, and I'd like you to get credit for it.

**Comment by onikienko** [comment]
Nothing to PR. Just a few lines of code. Tests did not pass since rollup hangs on `bundleImports`. So I am even not sure that these changes are correct. I just made sure that all path looks correct. If you want you may take a look at my changes https://github.com/onikienko/rollup-plugin-chrome-extension/commit/e5cd8a341a2244d80c49b1e0eea7ead7d9804734

**Comment by Dimirolz** [comment]
Hi @jacksteamdev, I had the same problem building project for Windows. At the same time on WSL it works great. 
I found the problem, and fix it for myself. It's dirty because I'm a newbie. And most likely my fix broke something else.

But maybe it gives you an idea and helps you to fix it in the right way. The problem in my case was here.

https://github.com/extend-chrome/rollup-plugin-chrome-extension/blob/1ac7985ce31dd93ce9c3d08b72ac76469b2f2ef1/src/files.machine.ts#L104-L118

I change these few lines of code:
```js
const dirName = process.cwd();
// and
manifest.id = dirName + "\\manifest.json";
```
And everything starts working on Windows.

**Comment by jacksteamdev** [comment]
@Dimirolz This helps a lot! It's good to know we don't have to change lots of things to get this working.

**Comment by jacksteamdev** [comment]
@onikienko @Dimirolz I've done a major rewrite in v4.0.1-16, would you be interested in giving this another look?

The API has changed a little, I've updated the instructions on this DEV article: 

[Create a Vite-React Chrome Extension in 90 seconds - DEV Community 👩‍💻👨‍💻](https://dev.to/jacksteamdev/create-a-vite-react-chrome-extension-in-90-seconds-3df7)

**Comment by DaPotatoMan** [comment]
@jacksteamdev With v4.0.1-16 `pnpm dev` works but fails to build.

```bash
> Executing task in folder demo: pnpm run build <


> demo@0.0.0 build X:\Projects\demo
> vite build

vite v2.8.4 building for production...
✓ 16 modules transformed.
rendering chunks (4)...[vite:build-html] The "fileName" or "name" properties of emitted files must be strings that are 
neither absolute nor relative paths, received "../../../X:\Projects\demo\index.html".
error during build:
Error: The "fileName" or "name" properties of emitted files must be strings that are neither absolute nor relative paths, received "../../../X:\Projects\demo\index.html".
    at error (X:\Projects\demo\node_modules\.pnpm\registry.npmjs.org+rollup@2.68.0\node_modules\rollup\dist\shared\rollup.js:159:30)
    at FileEmitter.emitFile (X:\Projects\demo\node_modules\.pnpm\registry.npmjs.org+rollup@2.68.0\node_modules\rollup\dist\shared\rollup.js:15476:24)
    at Object.generateBundle (X:\Projects\demo\node_modules\.pnpm\registry.npmjs.org+vite@2.8.4\node_modules\vite\dist\node\chunks\dep-971d9e33.js:21679:22)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
 ELIFECYCLE  Command failed with exit code 1.
The terminal process "C:\Program Files\PowerShell\7\pwsh.exe -Command pnpm run build" terminated with exit code: 1.
```
<br>
<details>
<summary>vite.config.ts</summary>

```typescript
import { defineConfig } from 'vite'
import vue from '@vitejs/plugin-vue'
import { chromeExtension } from "rollup-plugin-chrome-extension";
import manifest from './manifest.json'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [vue(), chromeExtension({ manifest })],
})
```
</details>

<details>
<summary>index.html</summary>

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <link rel="icon" href="/favicon.ico" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
  </head>
  <body>
    <div id="app"></div>
    <script type="module" src="/src/main.ts"></script>
  </body>
</html>
```
</details>

<details>
<summary>manifest.json</summary>

```json
{
  "manifest_version": 3,
  "name": "Rpce React Vite Example",
  "version": "1.0.0",
  "action": { "default_popup": "index.html" }
}
```
</details>

**Comment by pietrofxq** [comment]
I am having the same issue on Windows 10, can't build our chrome extension in my new machine :(

**Comment by keyeightysix** [comment]
I

> @jacksteamdev With v4.0.1-16 `pnpm dev` works but fails to build.
> 
> ```shell
> > Executing task in folder demo: pnpm run build <
> 
> 
> > demo@0.0.0 build X:\Projects\demo
> > vite build
> 
> vite v2.8.4 building for production...
> ✓ 16 modules transformed.
> rendering chunks (4)...[vite:build-html] The "fileName" or "name" properties of emitted files must be strings that are 
> neither absolute nor relative paths, received "../../../X:\Projects\demo\index.html".
> error during build:
> Error: The "fileName" or "name" properties of emitted files must be strings that are neither absolute nor relative paths, received "../../../X:\Projects\demo\index.html".
>     at error (X:\Projects\demo\node_modules\.pnpm\registry.npmjs.org+rollup@2.68.0\node_modules\rollup\dist\shared\rollup.js:159:30)
>     at FileEmitter.emitFile (X:\Projects\demo\node_modules\.pnpm\registry.npmjs.org+rollup@2.68.0\node_modules\rollup\dist\shared\rollup.js:15476:24)
>     at Object.generateBundle (X:\Projects\demo\node_modules\.pnpm\registry.npmjs.org+vite@2.8.4\node_modules\vite\dist\node\chunks\dep-971d9e33.js:21679:22)
>     at processTicksAndRejections (node:internal/process/task_queues:96:5)
>  ELIFECYCLE  Command failed with exit code 1.
> The terminal process "C:\Program Files\PowerShell\7\pwsh.exe -Command pnpm run build" terminated with exit code: 1.
> ```
> 
> 
> vite.config.ts
> index.html
> manifest.json

I also have the exact same output, on Windows and using the same files as this comment. I'm on latest version 4.0.1-16.

I'm totally plucking at strews here but looking at path.d.ts and searching basename within the plugin was where my rabbit hole stopped. It does look like everything is there to be normalising paths for windows but it's happening.

Not an issue from this project but looks related. https://github.com/eirslett/storybook-builder-vite/issues/31#issuecomment-867519691 

Windows path issue.

**Comment by jacksteamdev** [comment]
@DaPotatoMan @pietrofxq @keyeightysix Thanks for the feedback!

**Comment by keyeightysix** [comment]
> @DaPotatoMan @pietrofxq @keyeightysix Thanks for the feedback!

No problem at all @jacksteamdev. I was really surprised and happy to find this project only yesterday after a few weeks of extension land.

If there's anything I can do to help debug this issue let me know.

**Comment by Stumblinbear** [comment]
Also running into this issue--any luck on a workaround?

**Comment by lionelhorn** [comment]
``
rendering chunks (6)...[vite:build-html] The "fileName" or "name" properties of emitted files must be strings that are neither absolute nor relative paths, received "../../
../../../
``

I'm also expericiencing this issue. 
With version 4.0.1-18.
vite v2.8.6

**Comment by DaPotatoMan** [comment]
@Stumblinbear @lionelhorn There's a workaround if you modify a local vite file (in node_modules).
Try to search for this [line](https://github.com/vitejs/vite/blob/4194cce60f7d9a5664ef9aea279f1f69631cdc84/packages/vite/src/node/plugins/html.ts#L624) and change the code to:

```ts
this.emitFile({
  type: 'asset',
  fileName: path.basename(shortEmitName),
  source: result
})
```

It's an internal vite issue. I have created an issue [here](https://github.com/vitejs/vite/issues/7191).

**Comment by DaPotatoMan** [comment]
@jacksteamdev Looks like this issue is fixed by [#7383](https://github.com/vitejs/vite/pull/7383) in `Vite 2.9.0`. Both the `dev` and `build` command is working correctly.
